### PR TITLE
Implement custom element reactions

### LINKED ISSUE #17433: Implement Custom Element Reactions
spec: https://html.spec.whatwg.org/multipage/#custom-element-reactions

### PR DESCRIPTION
**Pull request by cbrewster**
Initial work for implementing custom element reactions: https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element-reactions

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17433 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17614)

### CODE REVIEW

**Comment by jdm** [comment]
```
failures:

    size_of::size_div

    size_of::size_element

    size_of::size_htmlelement

    size_of::size_span
```

**Comment by cbrewster** [comment]
Ready for review. r? @jdm 

I am going to submit the changes for `[CEReactions]` as a separate PR since it would add quite a lot of noise here. This means that all reactions are triggered via the backup element queue through a microtask. This is why many reaction-related tests are failing, they should be passing after `[CEReactions]` is implemented.

**Comment by nox** [comment]
The unit tests are still failing.

File: components/script/dom/customelementregistry.rs [hunk 1]
```diff
@@ -456,6 +456,196 @@ impl CustomElementDefinition {
...
+
+                    let namespace = DOMString::from(&*namespace);
+                    rooted!(in(cx) let mut namespace_value = UndefinedValue());
+                    unsafe { namespace.to_jsval(cx, namespace_value.handle_mut()); }
+
+                    let mut args = Vec::with_capacity(4);
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: `let mut args = vec![Heap::default(), Heap::default(), Heap::default(), Heap::default()];`

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -141,6 +143,9 @@ pub struct Element {
     /// when it has exclusive access to the element.
     #[ignore_heap_size_of = "bitflags defined in rust-selectors"]
     selector_flags: Cell<ElementSelectorFlags>,
+    custom_element_reaction_queue: DOMRefCell<VecDeque<CustomElementReaction>>,
+    #[ignore_heap_size_of = "Rc"]
+    custom_element_definition: DOMRefCell<Option<Rc<CustomElementDefinition>>>,
```
Comment (jdm) [inline components/script/dom/element.rs]: Add links to appropriate specification concepts?

File: components/script/dom/element.rs [hunk 3]
```diff
@@ -244,6 +249,8 @@ impl Element {
             class_list: Default::default(),
             state: Cell::new(state),
             selector_flags: Cell::new(ElementSelectorFlags::empty()),
+            custom_element_reaction_queue: Default::default(),
+            custom_element_definition: DOMRefCell::new(None),
```
Comment (jdm) [inline components/script/dom/element.rs]: Default::default() works here.

File: components/script/dom/customelementregistry.rs [hunk 4]
```diff
@@ -456,6 +456,196 @@ impl CustomElementDefinition {
...
+        ElementQueue {
+            queue: Default::default(),
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#invoke-custom-element-reactions
+    fn invoke_reactions(&self) {
+        // Steps 1-2
+        for element in self.queue.borrow().iter() {
+            element.invoke_reactions()
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: Can these reactions end up appending an element to this queue?

File: components/script/dom/element.rs [hunk 5]
```diff
@@ -278,6 +285,25 @@ impl Element {
...
+    }
+
+    pub fn push_reaction(&self, reaction: CustomElementReaction) {
+        self.custom_element_reaction_queue.borrow_mut().push_back(reaction);
+    }
+
+    pub fn invoke_reactions(&self) {
+        let mut reaction_queue = self.custom_element_reaction_queue.borrow_mut();
+        while let Some(reaction) = reaction_queue.pop_front() {
+            reaction.invoke(Root::from_ref(self));
```
Comment (jdm) [inline components/script/dom/element.rs]: Can these reactions end up appending reactions to the queue?

File: components/script/dom/node.rs [hunk 6]
```diff
@@ -313,6 +315,11 @@ impl Node {
             // e.g. when removing a <form>.
             vtable_for(&&*node).unbind_from_tree(&context);
             node.style_and_layout_data.get().map(|d| node.dispose(d));
+            if let Some(element) = node.downcast::<Element>() {
```
Comment (jdm) [inline components/script/dom/node.rs]: Add a link to https://dom.spec.whatwg.org/#concept-node-remove step 14 here.

File: components/script/dom/node.rs [hunk 7]
```diff
@@ -1401,13 +1408,22 @@ impl Node {
...
-            // Step 3.
+            // Step 3.1.
             for descendant in node.traverse_preorder() {
                 descendant.set_owner_doc(document);
             }
-            // Step 4.
             for descendant in node.traverse_preorder() {
+                // Step 3.2.
+                if let Some(element) = node.downcast::<Element>() {
+                    if element.get_custom_element_definition().is_some() {
```
Comment (jdm) [inline components/script/dom/node.rs]: We may want to write a `Node::as_custom_element` helper which combines the downcast and custom check into one.

File: tests/wpt/metadata/custom-elements/reaction-timing.html.ini [hunk 8]
```diff
@@ -1,5 +1,6 @@
 [reaction-timing.html]
   type: testharness
+  expected: CRASH
```
Comment (jdm) [inline tests/wpt/metadata/custom-elements/reaction-timing.html.ini]: ?

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/customelementregistry.rs (hunk 4 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: I do not believe so, there are two queues an element can be added to:
1. The current element queue.
2. The backup element queue.

For the first case, the current element queue is popped from the custom element reaction stack thus it is no longer the current element queue which means no element will be added to it.

The second case is the backup element queue has the `processing_backup_element_queue` flag which prevents elements being enqueued until the flag is set to `NotProcessing`.

File: components/script/dom/customelementregistry.rs [hunk 9]
```diff
@@ -456,6 +456,196 @@ impl CustomElementDefinition {
...
+
+                    let mut args = Vec::with_capacity(4);
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.get_mut(0).unwrap().set(name_value.get());
+                    args.get_mut(1).unwrap().set(old_value.get());
+                    args.get_mut(2).unwrap().set(value.get());
+                    args.get_mut(3).unwrap().set(namespace_value.get());
```
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Done.

File: components/script/dom/element.rs (hunk 3 above)
Comment (cbrewster) [inline components/script/dom/element.rs]: Done.

File: components/script/dom/element.rs (hunk 2 above)
Comment (cbrewster) [inline components/script/dom/element.rs]: Done.

File: components/script/dom/element.rs (hunk 5 above)
Comment (cbrewster) [inline components/script/dom/element.rs]: See other comment

File: components/script/dom/node.rs (hunk 7 above)
Comment (cbrewster) [inline components/script/dom/node.rs]: Done.

File: components/script/dom/customelementregistry.rs (hunk 4 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Oh actually in the second case, an element can be appended to the backup element queue is being invoked. The flag check occurs after the element is appended to the queue. This is what causes the crash in the test expectations.

I want to change it so the queue is not borrowed for the duration of the loop. Instead if element are popped and processed, new element can be appended to the queue. But I always run into unrooted must root errors. any ideas?

```
while let Some(element) = self.queue.borrow_mut().pop_front().map(|e| Root::from_ref(&*e)) {
    element.invoke_reactions()
}
```

Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Avoiding the closure appears to appease the lint
```
self.queue.borrow_mut().pop_front().as_ref().map(JS::deref).map(Root::from_ref)
```

File: components/script/dom/customelementregistry.rs [hunk 10]
```diff
@@ -330,6 +452,192 @@ impl CustomElementDefinition {
...
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#invoke-custom-element-reactions
+    fn invoke_reactions(&self) {
+        // Steps 1-2
+        while let Some(element) = self.next_element() {
+            element.invoke_reactions()
+        }
+        self.queue.borrow_mut().clear();
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: This shouldn't be necessary since we now pop, right?

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Looks like the sizeof unit tests are failing again, too.

**Pull request merged**